src/mvi/gpgmm_d3d12.h

```diff
@@ -64,7 +71,13 @@ namespace gpgmm::d3d12 {
       public:
         static HRESULT CreateHeap(const HEAP_DESC& descriptor,
                                   IResidencyManager* const pResidencyManager,
-                                  CreateHeapFn&& createHeapFn,
+                                  CreateHeapFnPtr createHeapFn,
```

Comment:
CreateHeap needs to also work with non-resources so passing resource-specific parameters like this won't work for other heap types. The call-back can only accept ID3D12Pageable as output which is what `CreateCommittedResourceCallback` should return and pass to CreateHeap.

Reply:
Any idea how to do that? Obviously I have to call CreateCommittedResource in a callback, which requires those resource-specific parameters. I have to get those to the callback somehow - but if I can't pass as parameters the only other option I can think of is some kind of binding model. 

Is it not an option to just pass a HeapParams struct that would contain a subset of possible heap creation parameters?

Reply:
(Sorry, didn't see the edit)

Reply:
D3D12 needs a lot of HeapParams (for. CreateDescriptorHeap, CreateQueryHeap, CreateFence, etc).

I was wondering if we could create `struct CreateResourceWrapper` which gets initialized to the HeapParams specific to the type. Then pass `CreateResourceWrapper::CreateFn` to CreateHeap which calls the D3D creation method and returns the ID3D12Pageable provided by CreateHeap.

Reply:
So `struct CreateResourceWrapper` would be a base and we'd create more structs that inherit from it i.e. `struct CreateCommittedResourceParams : CreateResourceWrapper` then we'd pass it through `CreateHeap`, pass it to the `CreateHeapFn`, then cast it to the appropriate type within `CreateHeapFn`?

Reply:
Pushed a new draft for the full gpgmm, which I think implements the idea described. Seems like it will work but this method means we can't pass in our CreateHeapFn in the exported API. It doesn't seem like there's a workaround unless I've misinterpreted the design suggestion. How much of a problem is this?

Reply:
It doesn't appear that the WebNN DML prototype actually calls the CreateHeap export...so we could just surround it with a NON_CHROMIUM_COMPLIANT ifdef and leave CreateHeapFn-based functionality - whereas the internal calls to CreateHeap could use the method from the current PR draft.

Reply:
Pushed another draft. Needs polish but lmk if the approach is as expected.